src/finalizer/finalizer.cpp

```diff
@@ -5,207 +5,325 @@
+                    // stores the ProductCode, DisplayName, and Version, but by calling into MsiGetProductInfo,
+                    // we're doing an implicit detect and getting a property back.
+                    hr = WiuGetProductInfo(sczProductId, L"ProductName", &sczProductName);
+                    if (SUCCEEDED(hr))
+                    {
+                        // The provider key *should* have the ProductName and ProductVersion properties, but since
+                        // we know it's installed, we just query the installer service. 
+                        MsiSetInternalUI(INSTALLUILEVEL_NONE, NULL);
+                        hr = WiuConfigureProductEx(sczProductId, INSTALLLEVEL_DEFAULT, INSTALLSTATE_ABSENT, L"MSIFASTINSTALL=7 IGNOREDEPENDENCIES=ALL REBOOT=ReallySuppress", &restart);
+                        LogStringLine(REPORT_STANDARD, "  Uninstall of \"%ls\" (%ls%) exited with 0x%.8x", sczProductName, sczProductId, hr);
+
+                        // Flag any reboot since we need to return that to the bundle.
+                        if (WIU_RESTART_INITIATED == restart || WIU_RESTART_REQUIRED == restart)
+                        {
+                            LogStringLine(REPORT_STANDARD, "  Reboot requested, deferring.");
+                            *pbRestartRequired = TRUE;
+                        }
+
+                        // Reset potential failures so we can continue to remove as many dependents as possible.
+                        hr = S_OK;
+                    }
+                    else if (HRESULT_FROM_WIN32(ERROR_UNKNOWN_PRODUCT) == hr || HRESULT_FROM_WIN32(ERROR_UNKNOWN_PROPERTY) == hr)
+                    {
                         // Possibly a corrupted provider key that wasn't cleaned up. We'll just ignore it.
-						LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
-						hr = S_OK;
-					}
-				}
-			}
-		}
+                        LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
+                        hr = S_OK;
+                    }
+                }
+            }
+        }
 
-		ReleaseRegKey(hkDependentsKey);
-		ReleaseRegKey(hkProviderKey);
-	}
+        ReleaseRegKey(hkDependentsKey);
+        ReleaseRegKey(hkProviderKey);
+    }
 
 LExit:
-	ReleaseStr(sczProductName);
-	ReleaseStr(sczProductId);
-	ReleaseStr(sczProviderKey);
-	ReleaseStr(sczDependentsKey);
-	ReleaseRegKey(hkDependentsKey);
-	ReleaseRegKey(hkProviderKey);
-	ReleaseRegKey(hkInstallerDependenciesKey);
-	return hr;
+    ReleaseStr(sczProductName);
+    ReleaseStr(sczProductId);
+    ReleaseStr(sczProviderKey);
+    ReleaseStr(sczDependentsKey);
+    ReleaseRegKey(hkDependentsKey);
+    ReleaseRegKey(hkProviderKey);
+    ReleaseRegKey(hkInstallerDependenciesKey);
+    return hr;
+}
+
+extern "C" HRESULT ParseSdkVersion(LPWSTR sczSdkVersion, INT * piMajor, INT * piMinor, INT * piFeatureBand)
```

Comment:
The naming here is confusing to me (maybe just because I'm not used to reading c++ :) )... is the 'pi' prefix because these are pointers? What about 'scz'? Also, are we expecting the feature band to be a full band number (ex 100) or just the first digit?

Reply:
yup, Hungarian notation, so pi is pointer to int. My C++ is very rusty. https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/efda8314-6e41-4837-8299-38ba0ee04b92 and https://docs.microsoft.com/en-us/windows/win32/stg/coding-style-conventions might helpt.

Yes for the feature band - similar to the managed SdkFeatureBand we have, so always 6.0.100 or 6.1.900